cpp/src/arrow/dataset/scanner.cc

```diff
@@ -708,8 +709,12 @@ Result<ProjectionDescr> ProjectionDescr::FromNames(std::vector<std::string> name
   for (size_t i = 0; i < exprs.size(); ++i) {
     exprs[i] = compute::field_ref(names[i]);
   }
+  auto fields = dataset_schema.fields();
+  for (const auto& aug_field : kAugmentedFields) {
+    fields.push_back(aug_field);
+  }
   return ProjectionDescr::FromExpressions(std::move(exprs), std::move(names),
-                                          dataset_schema);
+                                          Schema(fields));
```

Comment:
As it was asked to have the filename as a column, thus I have included it in the augmented fields as a scalar value, so that we can query for the column from the dataset. Can there be a better approach?

Reply:
We can then have a Python test to check if the correct filename exist in the column.

Reply:
It is indeed new behavior.  Before, we couldn't query any of the augmented fields.  We could make filename a special "queryable augmented field" since people are unlikely to want the others anytime soon but it seemed more straightforward just to make them all queryable.

Reply:
Ok, sounds reasonable to me.

Reply:
I think we should call this out in documentation in various places, then, else people may be surprised to see a set of new fields pop up out of nowhere.

Reply:
Agreed about the docs update.  And, actually, now that I think about it, we probably want to make sure that these fields are not included in the "default" projection (e.g. if both `projected_schema` and `projection` are not specified) and we should probably have a test case verifying that these fields do not get returned.

Reply:
Ah, the first test case in `TestScanner, ProjectionDefaults` would be the place for it.

Reply:
Will add the required docs for the change and the required test case.